### PR TITLE
Migrated country list to recycler view

### LINKED ISSUE #302: Migrate country list to RecyclerView
Subtask of #291

### PR DESCRIPTION
**Pull request by paetz**
Resolves #302

### CODE REVIEW

File: app/src/main/java/com/yacgroup/yacguide/UpdateHandler.kt [hunk 1]
```diff
@@ -18,7 +18,9 @@
 package com.yacgroup.yacguide
 
 import android.app.Dialog
+import android.view.View
 import android.widget.ImageButton
+import android.widget.LinearLayout
```
Comment (paetz) [inline app/src/main/java/com/yacgroup/yacguide/UpdateHandler.kt:23]: Not needed

File: app/src/main/java/com/yacgroup/yacguide/database/Country.kt [hunk 2]
```diff
@@ -18,9 +18,9 @@
...
 import androidx.room.Entity
+import androidx.room.PrimaryKey
 
-@Entity(primaryKeys = ["name"])
-class Country {
-
-    var name = ""
-}
+@Entity
+data class Country(
```
Comment (paetz) [inline app/src/main/java/com/yacgroup/yacguide/database/Country.kt:24]: FYI: `data` classes allow property-wise object comparison by implementing an according `==` operator. I use this feature in the adapter to check the difference of list items (which are updated by RecyclerView if they differ). Example: Returning to the route activity while its ascent bitmask has changed (and it hence should be colored).

File: app/src/main/res/layout/content_table.xml [hunk 3]
```diff
@@ -20,7 +20,7 @@
 <ScrollView
     xmlns:android="http://schemas.android.com/apk/res/android"
     android:layout_width="match_parent"
-    android:layout_height="match_parent"
+    android:layout_height="wrap_content"
```
Comment (paetz) [inline app/src/main/res/layout/content_table.xml:23]: ...otherwise there is no space left for the download button

File: app/src/main/res/layout/list_data_item.xml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+  -->
+<LinearLayout
```
Comment (paetz) [inline app/src/main/res/layout/list_data_item.xml:18]: Note: This item is generic and will be used for all object types from country down to route. That's why there are so many text fields.

File: app/src/main/res/layout/activity_country.xml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  Copyright (C) 2019 Fabian Kantereit
```
Comment (paetz) [inline app/src/main/res/layout/activity_country.xml:3]: Wrong name

File: app/src/main/java/com/yacgroup/yacguide/CountryActivity.kt [hunk 6]
```diff
@@ -36,35 +36,29 @@ class CountryActivity : TableActivityWithOptionsMenu() {
...
-            layout.addView(_updateHandler.getDownloadButton { displayContent() })
-        }
+        _viewAdapter.submitList(countries)
+        _updateHandler.configureDownloadButton(countries.isEmpty()){ displayContent() }
+    }
+
+    private fun _onCountrySelected(countryName: String) {
+        val intent = Intent(this@CountryActivity, RegionActivity::class.java)
+        intent.putExtra(IntentConstants.CLIMBING_OBJECT_LEVEL, ClimbingObjectLevel.eRegion.value)
+        intent.putExtra(IntentConstants.CLIMBING_OBJECT_PARENT_NAME, countryName)
```
Comment (chrgernoe) [inline app/src/main/java/com/yacgroup/yacguide/CountryActivity.kt:61]: Please use `apply`.

File: app/src/main/res/layout/button_download.xml [hunk 7]
```diff
@@ -17,13 +17,16 @@
 
 <ImageButton
     xmlns:android="http://schemas.android.com/apk/res/android"
-    style="@style/AppTheme.Secondary.ImageButton"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
```
Comment (chrgernoe) [inline app/src/main/res/layout/button_download.xml:20]: Why did you change the button theme?

File: app/src/main/res/layout/list_view_content.xml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!--
+  Copyright (C) 2019 Axel Paetzold
```
Comment (chrgernoe) [inline app/src/main/res/layout/list_view_content.xml:4]: Wrong year.

File: app/src/main/res/values/colors.xml [hunk 9]
```diff
@@ -31,8 +31,8 @@
     <color name="colorOnPrimaryDark">@color/white</color>
 
     <!-- NOTE: Must be in sync with the colors in the WidgetUtils class. -->
-    <color name="colorSecondary">#9e9e9e</color>
-    <color name="colorSecondaryLight">#cfcfcf</color>
```
Comment (chrgernoe) [inline app/src/main/res/values/colors.xml:35]: Why did you change the color?

**Review by chrgernoe — CHANGES_REQUESTED** [review]
The list items are now have a very high height. I looks a bit odd for the country names.

**Comment by paetz** [comment]
> The list items are now have a very high height. I looks a bit odd for the country names.

You really think so? Whereas I find the old (=current) country appearance quite odd since it's much too small.

File: app/src/main/res/values/colors.xml (hunk 9 above)
Comment (paetz) [reply app/src/main/res/values/colors.xml:35]: Because I added margin to the layout. Doing so, the old background color didn't really fit it.

File: app/src/main/res/layout/button_download.xml (hunk 7 above)
Comment (paetz) [reply app/src/main/res/layout/button_download.xml:20]: That's what I already disliked about the old impl., that the download wasn't really visualized like a button. This way, I find it quite easier for the user.

**Comment by chrgernoe** [comment]
> > The list items are now have a very high height. I looks a bit odd for the country names.
> 
> You really think so? Whereas I find the old (=current) country appearance quite odd since it's much too small for my fingers. Compromise: I'm gonna make the info text field on top invisible by default. So that item height is only 2 * currHeight instead of 3 * currHeight

Now, it looks OK. There is another small thing: Why is there a separator on top but no additional item?

File: app/src/main/res/layout/button_download.xml (hunk 7 above)
Comment (chrgernoe) [reply app/src/main/res/layout/button_download.xml:20]: Please do not change the style individual for each item. Use styles instead. Otherwise it gets a mess again. I tried to clean this up with PR #217.

File: app/src/main/res/values/colors.xml (hunk 9 above)
Comment (chrgernoe) [reply app/src/main/res/values/colors.xml:35]: OK

**Comment by chrgernoe** [comment]
In general, please have a look here: https://www.gnu.org/licenses/gpl-howto.en.html

**Review by chrgernoe — CHANGES_REQUESTED** [review]

**Comment by paetz** [comment]
> > > The list items are now have a very high height. I looks a bit odd for the country names.
> > 
> > 
> > You really think so? Whereas I find the old (=current) country appearance quite odd since it's much too small for my fingers. Compromise: I'm gonna make the info text field on top invisible by default. So that item height is only 2 * currHeight instead of 3 * currHeight
> 
> Now, it looks OK. There is another small thing: Why is there a separator on top but no additional item?

This is by intent to visualize the table start since there may be other text fields above like rock status.

**Review by chrgernoe — APPROVED** [review]
